keps/sig-etcd/4743-kuberernetes-etcd-interface/README.md

```diff
@@ -0,0 +1,351 @@
+[Prober]: https://github.com/kubernetes/kubernetes/blob/6ba9fa89fb5889550649bfde847c742a55d3d29c/staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go#L256-L268
+[Feature checker]: https://github.com/kubernetes/kubernetes/blob/6ba9fa89fb5889550649bfde847c742a55d3d29c/staging/src/k8s.io/apiserver/pkg/storage/feature/feature_support_checker.go#L143-L151
+[Design Doc: etcd livez and readyz probes]: https://docs.google.com/document/d/1SkzmO4RT_GI9YhT0dw4a6nEwKVCciwrwbCDxK0D7ASM/edit?usp=sharing
+[KEP-4647]: https://github.com/kubernetes/enhancements/pull/4662
+[Lease manager]: https://github.com/kubernetes/kubernetes/blob/6ba9fa89fb5889550649bfde847c742a55d3d29c/staging/src/k8s.io/apiserver/pkg/storage/etcd3/lease_manager.go#L90-L120
+[#110210]: https://github.com/kubernetes/kubernetes/issues/110210
+
+## Proposal
+
+This KEP proposes creating an etcd-Kubernetes code interface owned and
+maintained by SIG-etcd. The interface will serve as a formalization of the
+existing etcd-Kubernetes contract, ensuring the correct usage of etcd within
+Kubernetes and enabling improved testing and validation.
+
+The interface will prioritize etcd's existing capabilities and behaviors,
+focusing on compatibility with the current etcd API. It will not introduce
+features or behaviors not supported by etcd, adhering to the existing
+SIG API Machinery policy outlined in [Storage for Extension API Servers].
+This policy designates etcd as the sole supported storage backend for Kubernetes
+for the foreseeable future.
+
+[Storage for Extension API Servers]: https://docs.google.com/document/d/1i0xzRFB-uGLmLYueLMBTpHrOot9ScFxpkkcVcZHVbyA/edit?usp=sharing]
+
+## User Stories
+
+To better understand the importance of code location let’s visit the following use cases:
+
+### Kubernetes Backporting an etcd client patch version
+
+**The Journey:** Kubernetes regularly updates to newer etcd versions to leverage
+bug fixes, or security patches. However, ensuring compatibility between the 
+codified etcd-Kubernetes interface and etcd client is essential.
+
+**Considerations:**
+
+*   Even minor etcd client updates might inadvertently introduce changes that 
+    break the interface's assumptions or functionality.
+*   Kubernetes needs to choose a compatible interface version before upgrading 
+    the etcd client.
+*   Tight coupling to the etcd client could necessitate backporting the 
+    interface to older etcd branches, a complex and time-consuming process.
+
+
+### Making Changes/Patching Bugs in the Interface
+
+**The Journey:** Despite careful design, the complex etcd-Kubernetes contract 
+might reveal bugs or require adjustments.
+
+**Considerations:**
+
+*   Changes and bug fixes need to be implemented and released with minimal
+    disruption to both Kubernetes and etcd users.
+*   Tightly coupled interfaces (Options 1 and 2) might require a full etcd
```

Comment:
`(Options 1 and 2)` -> `(see alternatives section)`

Reply:
Done